value/value.go

```diff
@@ -319,96 +502,68 @@ func (m *Map) Less(rhs *Map) bool {
+	Delete(string)
+	Iterate(func(string, Value) bool)
 }
 
-// Set inserts or updates the given item.
-func (m *Map) Set(key string, value Value) {
-	if f, ok := m.Get(key); ok {
-		f.Value = value
-		return
-	}
-	m.Items = append(m.Items, Field{Name: key, Value: value})
-	i := len(m.Items) - 1
-	m.index[key] = i
-	m.order = nil
-}
-
-// Delete removes the key from the set.
-func (m *Map) Delete(key string) {
-	items := []Field{}
-	for i := range m.Items {
-		if m.Items[i].Name != key {
-			items = append(items, m.Items[i])
+type MapString map[string]interface{}
+
+func (m MapString) Length() int {
+	return len(m)
+}
+
+func (m MapString) Get(key string) (Value, bool) {
+	val, ok := m[key]
+	return ValueInterface{Value: val}, ok
+}
+
+func (m MapString) Set(key string, val Value) {
+	m[key] = val.Interface()
+}
+
+func (m MapString) Delete(key string) {
+	delete(m, key)
+}
+
+func (m MapString) Iterate(iter func(string, Value) bool) {
+	for key, value := range m {
+		if !iter(key, ValueInterface{Value: value}) {
+			return
 		}
 	}
-	m.Items = items
-	m.index = nil // Since the list has changed
-	m.order = nil
 }
 
-// StringValue returns s as a scalar string Value.
-func StringValue(s string) Value {
-	s2 := String(s)
-	return Value{StringValue: &s2}
+type MapInterface map[interface{}]interface{}
+
+func (m MapInterface) Length() int {
+	return len(m)
 }
 
-// IntValue returns i as a scalar numeric (integer) Value.
-func IntValue(i int) Value {
-	i2 := Int(i)
-	return Value{IntValue: &i2}
+func (m MapInterface) Get(key string) (Value, bool) {
+	val, ok := m[key]
+	return ValueInterface{Value: val}, ok
 }
 
-// FloatValue returns f as a scalar numeric (float) Value.
-func FloatValue(f float64) Value {
-	f2 := Float(f)
-	return Value{FloatValue: &f2}
+func (m MapInterface) Set(key string, val Value) {
+	m[key] = val.Interface()
 }
 
-// BooleanValue returns b as a scalar boolean Value.
-func BooleanValue(b bool) Value {
-	b2 := Boolean(b)
-	return Value{BooleanValue: &b2}
+func (m MapInterface) Delete(key string) {
+	delete(m, key)
 }
 
-// String returns a human-readable representation of the value.
-func (v Value) String() string {
-	switch {
-	case v.FloatValue != nil:
-		return fmt.Sprintf("%v", *v.FloatValue)
-	case v.IntValue != nil:
-		return fmt.Sprintf("%v", *v.IntValue)
-	case v.StringValue != nil:
-		return fmt.Sprintf("%q", *v.StringValue)
-	case v.BooleanValue != nil:
-		return fmt.Sprintf("%v", *v.BooleanValue)
-	case v.ListValue != nil:
-		strs := []string{}
-		for _, item := range v.ListValue.Items {
-			strs = append(strs, item.String())
+func (m MapInterface) Iterate(iter func(string, Value) bool) {
+	for key, value := range m {
+		vk, ok := key.(string)
+		if !ok {
+			continue
 		}
-		return "[" + strings.Join(strs, ",") + "]"
-	case v.MapValue != nil:
-		strs := []string{}
-		for _, i := range v.MapValue.Items {
-			strs = append(strs, fmt.Sprintf("%v=%v", i.Name, i.Value))
+		if !iter(vk, ValueInterface{Value: value}) {
```

Comment:
This is the thing that escapes?

Reply:
yes

Reply:
```
value/value.go:565:30: ValueInterface literal escapes to heap
value/value.go:565:30:  from iter(vk, ValueInterface literal) (parameter to indirect call) at value/value.go:565:11
```

Reply:
Here's an idea, probably worth what you paid for it:

```
type mapValue struct {
  m map[string]interface{}
  k string
}

...

if !iter(vk, mapValue{m, vk}) {
...
```

Reply:
but the mapValue is still going to escape?

Reply:
mapValue is not a pointer though...

Reply:
`ValueInterface{}` is not a pointer either?

Reply:
oh wait, it gets moved to the heap when stuffed in an interface? :(

I guess that makes sense.

Reply:
Yeah ...

Reply:
If that's the problem, we can allocate a single one and use it for everything?

Reply:
I would have to think of the consequences of doing so? Are we sure nothing is going to keep a copy of it somewhere?

Reply:
Hopefully that would break a test...

Reply:
I'm pretty sure it would, we can try this out